Wrappers/Python/cil/optimisation/utilities/preconditioner.py

```diff
@@ -0,0 +1,281 @@
+                self.array, where=self.array.as_array() > 0, out=self.array)
+
+    def apply(self, algorithm, gradient, out=None):
+        r"""
+        Update the preconditioner.
+
+        Parameters
+        ----------
+        algorithm : object
+            The algorithm object.
+        gradient : DataContainer
+            The calculated gradient to modify
+        out : DataContainer, optional, default is None 
+            The modified gradient to return 
+        """
+        if out is None:
+            out = gradient.copy()
+        gradient.multiply(
+            self.array, out=out)
+        return out
+
+
+class AdaptiveSensitivity(Sensitivity):
+
+    r"""
+    Adaptive Sensitivity preconditioner class. 
+
+    In each call to the preconditioner the `gradient` is multiplied by :math:` (x+\delta) \odot y /(A^T \mathbf{1})` where :math:`A` is an operator,  :math:`\mathbf{1}` is an object in the range of the operator filled with ones and :math:`y` is an optional reference image.  
+    The point :math:`x` is the current iteration and :math:`delta` is a small positive float. 
+
+
+    Parameters
+    ----------
+    operator : CIL object
+        The operator used for sensitivity computation.
+    delta : float, optional
+        The delta value for the preconditioner.
+    iterations : int, optional
+        The maximum number of iterations before the preconditoner is frozen and no-longer updates. Freezing iterations saves computational effort. 
+    reference : DataContainer e.g. ImageData, optional
+        The reference data, an object in the domain of the operator. Recommended to be a best guess reconstruction. 
+
+    Reference
+    ---------
+    Twyman R, Arridge S, Kereta Z, Jin B, Brusaferri L, Ahn S, Stearns CW, Hutton BF, Burger IA, Kotasidis F, Thielemans K. An Investigation of Stochastic Variance Reduction Algorithms for Relative Difference Penalized 3D PET Image Reconstruction. IEEE Trans Med Imaging. 2023 Jan;42(1):29-41. doi: 10.1109/TMI.2022.3203237. Epub 2022 Dec 29. PMID: 36044488.
+
+    """
+
+    def __init__(self, operator, delta=1e-6, iterations=100, reference=None):
+
+        self.iterations = iterations
+        self.delta = delta
+        self.freezing_point = operator.domain_geometry().allocate(0)
+
+        super(AdaptiveSensitivity, self).__init__(
+            operator=operator, reference=reference)
+
+    def apply(self, algorithm, gradient, out=None):
+        r"""
+        Update the preconditioner.
+
+        Parameters
+        ----------
+        algorithm : object
+            The algorithm object.
+        gradient : DataContainer
+            The calculated gradient to modify
+        out : DataContainer, optional, default is None 
+            The modified gradient to return 
+        """
+        if out is None:
+            out = gradient.copy()
+            
+        if algorithm.iteration <= self.iterations:
+            self.array.multiply(algorithm.x + self.delta,
```

Comment:
It is safer to use `algorithm.get_output()` instead of `algorithm.x` unless you are sure that all algorithms that will use this preconditioner have `.x` attribute.

Reply:
Good shout - thank-you!